### PR TITLE
Add PowerLaw model and n_pred calculation

### PR DESCRIPTION
**Pull request by joleroi**
This PR does the following changes:

* [x] Add utility to get n_pred for a ``SpectrumObservation``
* [x] Copy some of the test from ``examples`` to real tests

cc @jjlk 

Remaining TODO for a future PR:
- Make energy binning optional (at the moment its hardcoded to what comes from the IRFS

### CODE REVIEW

File: examples/test_spectrum_model.py [hunk 1]
```diff
@@ -1,47 +1,74 @@
...
     SpectralModel, PowerLaw, ExponentialCutoffPowerLaw
 )
+from gammapy.irf import (
+    EffectiveAreaTable,
+    EnergyDispersion
+)
+from gammapy.utils.energy import EnergyBounds
+
+
+# TODO clean up and add to gammapy.irf
```
Comment (cdeil) [inline examples/test_spectrum_model.py]: Did you see this?
https://gammapy.readthedocs.io/en/latest/api/gammapy.irf.EnergyDispersion.html#gammapy.irf.EnergyDispersion.from_gauss

Maybe you could check / fix that Gaussian energy resolution and add this "perfect" resolution as a separate similar method?
I'm never quite sure where to put such "make example something" functions ... but maybe classmethods are best?

File: gammapy/spectrum/models.py [hunk 2]
```diff
@@ -2,24 +2,90 @@
...
 
 class SpectralModel(object):
     """Spectral model base class.
     """
+    def __call__(self, energy):
+        kwargs = self.parameters
+        kwargs.update(energy=energy)
+        return self.evaluate(**kwargs)
+
+    def predicted_counts(self, livetime, aeff, edisp):
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: I don't think the spectral model class is a good location for `predicted_counts`.

How about putting it here, taking a spectral model as input (i.e. not needing an import)?
http://docs.gammapy.org/en/latest/api/gammapy.data.DataStoreObservation.html

The advantage is that this `Observation` class has access to IRFs and livetime, and the livetime / GTI handling will get a little more complicated, so it's better to have the `npred` code there I think.

File: gammapy/spectrum/models.py [hunk 3]
```diff
@@ -2,24 +2,90 @@
...
+        counts = edisp.apply(counts.decompose())
+        return CountsSpectrum(data=counts, energy=edisp.reco_energy)
+        
 
 class PowerLaw(SpectralModel):
     """Spectral power-law model.
+    
+    .. math:: 
+
+        F(E) = F_0 \\cdot \\left( \\frac{E}{E_0} \\right)^{\Gamma}
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: Please use "raw" literal docstrings, i.e. add an `r"the docstring"` prefix, and then you can remove the double backslashes from the LaTeX formula, i.e. have a normal LaTeX string.

File: examples/test_spectrum_model.py (hunk 1 above)
Comment (joleroi) [inline examples/test_spectrum_model.py]: Yes, I did. but the ``from_gauss`` methods also needs to be tested/extended to nxm and I wanted to do that in a separate PR.

File: gammapy/spectrum/models.py (hunk 2 above)
Comment (joleroi) [inline gammapy/spectrum/models.py]: I agree that an observation class intuitively is the better place. However, the [SpectrumObservation](http://docs.gammapy.org/en/latest/api/gammapy.spectrum.SpectrumObservation.html) is a better place, since I already has accees to the 1D IRFs. The drwaback of putting the npred method here is, that people like @jjlk who only want to predict counts without having measuread ``n_on`` and ``n_off`` then cannot use this method. So I would say this is a show stopper. 
I will make a function 
```
def predicted_counts(model, arf, livetime, rmf)
```

and add a convenience method to the ``SpectrumObservation``, ok?

```

Comment (cdeil) [inline gammapy/spectrum/models.py]: @joleroi - Yes, sounds good.

Alternatively, the counts `n_on` and `n_off` could be optional (default: None) on SpectrumObservation construction.
And then @jjlk could also just create a `SpectrumObservation` to compute `npred`.

Also .. why does the [SpectrumObservation](http://docs.gammapy.org/en/latest/api/gammapy.spectrum.SpectrumObservation.html) `__init__` not take `livetime`?

I see the attribute is called `exposure` ... I'd prefer `livetime`, `exposure` sometimes means `A * T` and could be confusing.

Comment (joleroi) [inline gammapy/spectrum/models.py]: I decided to be consistent with ``OGIP`` keywords where exposure means livetime.

 Also, all meta information is stored on the ``on_vector`` = PHA file, since this is where it is stored on disk. It was different in the past, but then the ``meta`` data was scattered all over the place and got out of sync all the time.

**Comment by cdeil** [comment]
@joleroi - Unfortunately for you travis-ci didn't run yet and now I had time to have a look at this. :-)

I'll leave some comments inline, but my main one is this:

So `calculate_predicted_counts` returns a `CountSpectrum`?
Why not make it a classmethod on `CountSpectrum, then?
At the very least document in the docstring that this function returns a `CountSpectrum` object.

File: examples/test_spectrum_model.py [hunk 4]
```diff
@@ -1,47 +1,79 @@
...
 )
+from gammapy.irf import (
+    EffectiveAreaTable,
+    EnergyDispersion
+)
+from gammapy.utils.energy import EnergyBounds
+from gammapy.spectrum import calculate_predicted_counts
+
+
+# TODO clean up and add to gammapy.irf
```
Comment (cdeil) [inline examples/test_spectrum_model.py]: Why not move it to an `EnergyDispersion` classmethod now?
It's not a lot of work and then it's directly in the right place and easier to find and use.

File: gammapy/spectrum/core.py [hunk 5]
```diff
@@ -427,6 +396,20 @@ def stats(self, nbin):
             livetime = self.exposure,
         )
         return ObservationStats(**kwargs)
+    
+    def predicted_counts(self, model):
```
Comment (cdeil) [inline gammapy/spectrum/core.py]: Document what this method returns, i.e. a `CountSpectrum`.

Rename to `n_pred`?

File: gammapy/spectrum/models.py [hunk 6]
```diff
@@ -2,24 +2,66 @@
...
     'PowerLaw',
     'ExponentialCutoffPowerLaw',
 ]
 
+# Note: Consider to move stuff from _models_old.py here
 
 class SpectralModel(object):
     """Spectral model base class.
     """
+    def __call__(self, energy):
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: Change implementation to this:
```
def __call__(self, energy):
    return self.evaluate(energy, **self.parameters)
```

I think your current implementation adds `energy` to `self.parameters` and this is not intended?!
(Try it ... I'm pretty sure from looking at the code.)

Comment (cdeil) [inline gammapy/spectrum/models.py]: Maybe add a docstring for `__call__` with one sentence what it is?

File: gammapy/spectrum/models.py [hunk 7]
```diff
@@ -2,24 +2,66 @@
...
+    ----------
+    index : float, `~astropy.units.Quantity`
+        :math:`\Gamma`
+    amplitude: float, `~astropy.units.Quantity` 
+        :math:`F_0`
+    reference : float, `~astropy.units.Quantity` 
+        :math:`E_0`
     """
+    def __init__(self, index, amplitude, reference):
+        self.parameters = Bunch(index = index,
```
Comment (cdeil) [inline gammapy/spectrum/models.py]: I think I like the idea of storing `parameters` in a `Bunch`.
Maybe rename to `pars`? (either way is fine with me)

Can you mention this somewhere, maybe the docstring of the base class?

File: gammapy/spectrum/tests/test_spectrum_fit.py [hunk 8]
```diff
@@ -30,3 +30,13 @@ def test_spectral_fit():
     assert_allclose(fit.result.fit.statval, 100.433, rtol=1e-3)
     assert_quantity_allclose(fit.result.fit.parameters.index,
                              2.405 * u.Unit(''), rtol=1e-3)
+    
+    # Compare Npred vectors
+    # TODO: Read model from FitResult
+    from gammapy.spectrum.models import PowerLaw
```
Comment (cdeil) [inline gammapy/spectrum/tests/test_spectrum_fit.py]: Move import to the top of the file, and use relative imports within Gammapy.

File: gammapy/spectrum/tests/test_spectrum_fit.py [hunk 9]
```diff
@@ -30,3 +30,13 @@ def test_spectral_fit():
...
+    model = PowerLaw(index=fit.result.fit.parameters.index,
+                     reference=fit.result.fit.parameters.reference,
+                     amplitude=fit.result.fit.parameters.norm
+                    )
+    npred = obs1.predicted_counts(model)
+    assert_allclose(fit.result.fit.n_pred, npred.data, rtol=1e-3)
```
Comment (cdeil) [inline gammapy/spectrum/tests/test_spectrum_fit.py]: Nice test!

File: gammapy/spectrum/utils.py [hunk 10]
```diff
@@ -94,63 +94,36 @@ def bin_edges(self, energy):
...
-        Directory holding npred fits files
-    ogip_dir : str, Path
-        Directory holding OGIP data
+    model : `~gammapy.spectrum.models.SpectralModel`
+        Spectral model
+    livetime : `~astropy.units.Quantity`
+        Observation duration
+    aeff : `~gammapy.irf.EffectiveAreaTable`
+        EffectiveArea
+    edisp : `~gammapy.irf.EnergyDispersion`, optional
```
Comment (cdeil) [inline gammapy/spectrum/utils.py]: Here you're saying `edisp` is optional.
I don't think the function currently works without `edisp`!?

Maybe fix this and add a test?
(or leave a TODO for later)

File: gammapy/spectrum/utils.py [hunk 11]
```diff
@@ -94,63 +94,36 @@ def bin_edges(self, energy):
...
-    ax.legend(numpoints=1)
-    plt.xscale('log')
-
-    return ax
+    from . import CountsSpectrum
 
+    true_energy = aeff.energy.data
+    flux = model.integral(emin=true_energy[:-1], emax=true_energy[1:]) 
+    counts = flux * livetime * aeff.evaluate()
+    counts = counts.decompose()
```
Comment (cdeil) [inline gammapy/spectrum/utils.py]: I find this `decompose` stuff a bit weird and complicated to look at.
Maybe make it an array via `counts = counts.to('').value`?
(Not sure this is better)

File: examples/test_spectrum_model.py (hunk 4 above)
Comment (joleroi) [inline examples/test_spectrum_model.py]: We agreed to not blow up this PR and do this together with ``from_gauss`` in a follow-up PR

File: gammapy/spectrum/core.py (hunk 5 above)
Comment (joleroi) [inline gammapy/spectrum/core.py]: explicit > jargon

File: gammapy/spectrum/models.py (hunk 6 above)
Comment (joleroi) [inline gammapy/spectrum/models.py]: :+1:

File: gammapy/spectrum/tests/test_spectrum_fit.py (hunk 8 above)
Comment (joleroi) [inline gammapy/spectrum/tests/test_spectrum_fit.py]: This is a temp solution (see TODO)

File: gammapy/spectrum/utils.py (hunk 11 above)
Comment (joleroi) [inline gammapy/spectrum/utils.py]: I don't think ``counts.to('').value`` is easier to read thatn ``counts.decompose()``

**Comment by joleroi** [comment]
> So calculate_predicted_counts returns a CountSpectrum?
Why not make it a classmethod on CountSpectrum, then?

In the last implementation it was a classmethod in CountsSpectrum and you complained that it was a strange place ...

I don't think it's a good idea to put everything in classmethods. If I wanted to compute ``n_pred`` it would never come to my mind that I have to 
```import CountsSpectrum

n_pred = CountsSpectrum.n_pred(...)
```

**Comment by joleroi** [comment]
Remaining failures are due to setup.py magic

@cdeil  I either adressed your comments or explained why I refuse to implement them :wink: 

So I am merging this. Everything stays up to discussion as usual.
